doc/src/errorrecovery.md

```diff
@@ -1,8 +1,396 @@
+    | Factor { $1 }
+    ;
+
+Factor: 'LBRACK' Expr 'RBRACK' { $2 }
+      | 'INT' { parse_int($lexer.lexeme_str(&$1.unwrap())) }
+      ;
+%%
+// Functions / imports in this section are in scope for
+// each productions' actions.
+
+fn parse_int(s: &str) -> u64 {
+    match s.parse::<u64>() {
+        Ok(val) => val as u64,
+        Err(_) => panic!("{} cannot be represented as a u64", s)
+    }
+}
+```
+
+In this grammar, every production has an action: each action *must* evaluate to
+an instance of the `%type` type (in this case `u64`). The `$x` variables refer
+to the respective symbol in the production (i.e. `$1` refers to the first symbol
+in the production). If the symbol is a rule then an instance of `%type` is
+stored in the `$x` variable; if the symbol is a lexeme then an `Option<Lexeme>`
+instance is returned. A special `$lexer` variable allows access to the lexer.
+This allows us to turn `Lexeme`s into strings with the `lexeme_str` function,
+which, given a `Lexeme`, returns a `&str` of the relevant part of the user’s
+input.
+
+For many examples, this simple grammar and its actions work well leading
+to output such as the following:
+
+```
+>>> 2 + + 3
+Parsing error at line 1 column 5. Repair sequences found:
+   1: Delete +
+   2: Insert INT
+Result: 5
+```
+
+`Insert x` means “error recovery inserted a lexeme of type *x*”; `Delete x`
+means “error recovery deleted the next lexeme in the stream”; and `Shift x`
+means “error recovery kept the user’s lexeme *x* as-is”.
+
+Repair sequences are minimal ways of adjusting the user’s input such that it
+becomes correct relative to the underlying grammar. Intuitively, in this
+example, the two repair sequences would adjust the input to be equivalent to
+`2 + 3` (repair sequence 1) or `2 + <some int> + 3` (repair sequence 2). When
+more than one repair sequence is presented to the user, the first is used by the
+algorithm to continue parsing: in this case, the input was parsed as if it was
+equivalent to `2 + 3`, hence the evaluation of the input to `5`.
+
+Repair sequences can, as their name suggests, be of arbitrary length:
+
+```
+>>> 2 + 3 4 5
+Parsing error at line 1 column 7. Repair sequences found:
+   1: Insert MUL, Delete 4
+   2: Insert PLUS, Delete 4
+   3: Delete 4, Delete 5
+   4: Insert MUL, Shift 4, Delete 5
+   5: Insert MUL, Shift 4, Insert PLUS
+   6: Insert MUL, Shift 4, Insert MUL
+   7: Insert PLUS, Shift 4, Delete 5
+   8: Insert PLUS, Shift 4, Insert PLUS
+   9: Insert PLUS, Shift 4, Insert MUL
+Result: 17
+```
+
+In this case, the first repair sequence caused the input to be parsed as if it
+was equivalent to `2 + 3 * 5`, hence the evaluation of the input to `17`.
+
+
+## Syntax errors and language semantics
+
+Our example inputs so far have deliberately exploited cases where the first
+repair sequence at worst inserted “unimportant” lexemes such as `+` and `*`.
+Since the grammar’s actions never read the contents of such lexemes, only their
```

Comment:
I'm wondering if "value" is a better word to describe a lexeme's contents.

Reply:
Agreed.